### ISSUE TITLE
Improving the way how text selection works

### ISSUE THREAD

**Issue by mitar** [mozilla/pdf.js#4843]
For our project [PeerLibrary](http://peerlibrary.org/) we improved the way how text selection works in PDF.js. We are providing annotations on top of publications and the issue we were having was that in current version of PDF.js one has to be really precise when selecting a text and not move mouse into the empty space between text because selection jumps to the beginning. So annotation was really hard. We fixed so that it works. It is now much more intuitive in the sense that you can start selecting and in the middle of the text and as you get over the edge, selection gets to the end of the text and stays there.

Try it here: https://peerlibrary.org/p/tfgbf3kfxbQ4ubunj/trace-based-just-in-time-type-specialization-for-dynamic-languages

I am writing here because we would like to get this upstream into pdf.js, but are unsure if you will accept it, because a bit of hackery is needed and there are some edge cases. (Source code for our viewer is [here](https://github.com/peerlibrary/peerlibrary/blob/development/client/lib/highlighter.coffee). It would have to be ported to pdf.js viewer.)

The basic idea how it works is simple: as you drag a selection, we dynamically modify padding of the closest text segment element so that it always stays under the mouse. This prevents mouse dragging going to the parent canvas element and thus jumping of the selection. We were using it for some time now and it works very well.

There is a drawback. Now it is really important that text segments are in the text flow order because we use order of text segments to know how padding should behave. Sadly, this is not the case for all PDFs. We observed that sometimes footer and header elements are out of the flow. Sometimes watermarks are added in post-processing and they are out of the flow as well. In such cases we pad the wrong text segment which then prevents selection of the real content you want to select.

Would this be something you would like to have in pdf.js and should we spend the effort of porting it to pdf.js viewer? To really make it work in all PDFs, text layer segments should be ordered in the correct way. We do not have that code so that is something which should be done in parallel and do not really know how to do it properly.

In fact, correcting the text layer segments order would also make sure that when joining all text segments you get the correct text content of the PDF file.

(I have not yet checked latest text extraction improvements I see you made in the past few weeks and maybe this is fixed there already?)

**Comment by scharf** [comment]
What about a totally different approach to text selection: suppose you would add a 'text selection' mode: in that mode, you would show the hidden `textLayer` `div` elements. The the text could be selected pretty easily. You would loose some of the pixel perfect alignment of the text.

For example, I have taken the [compressed_tracemonkey-pldi-09_pdf](http://mozilla.github.io/pdf.js/web/viewer.html) example and made `.textLayer > div { color: red;}`. This reveals the hidden text layer:

![compressed_tracemonkey-pldi-09_pdf](https://cloud.githubusercontent.com/assets/3731059/6517910/470501e0-c3a0-11e4-9f03-1ca09e67f1b0.jpg)

As you can see the hidden text layer is slightly distorted and although it looks as if I had selected `traces` I actually have selected `aces,`.

So, the idea would be: show the users the hidden text layer and let them select that. That way they see what they get. They may also see that some parts of the document are not selectable.

One might want to show a dimmed version of the rendered PDF in the background, so that the user can see where some text appears in images.

What do you think?

**Comment by speedplane** [comment]
@scharf Some challenges with this approach:
- Selecting Would be Harder: users  would need to go into a special mode just to select text.
- It would be more confusing: For most users, it would just look like blurry text, they wouldn't be able to tell that there is a text layer that is slightly off.
- It's different: no other PDF viewer works that way, PDF.js should be a drop-in replacement for the others.
This issue is fixed by PR #5545, I'd vote to go that route.

**Comment by scharf** [comment]
The solution would be a 'hack' in case a real solution is not doable. When the alternative is to not being able to select correctly, then using this hack could be a step forward. 

Here is a little snippet to drop into the console, to play with this approach:

```javascript
ed = (function() {
   // hack: add jquery to the page
    var script = document.createElement("script");
    script.src = "//ajax.googleapis.com/ajax/libs/jquery/1/jquery.min.js";
    document.head.appendChild(script);
    var inedit = false;

    function toggle() {
        if (inedit) {
            inedit = false;
            jQuery('canvas').css('opacity', '');
            jQuery('.textLayer').css('opacity', '');
            jQuery('.textLayer > div').css('color', '');
        } else {
            inedit = true;
            jQuery('canvas').css('opacity','0.2');
            jQuery('.textLayer').css('opacity','0.8');
            jQuery('.textLayer > div').css('color','black');
        }
    }
    return toggle;
})()
```
The selection color also needs to change.

![compressed_tracemonkey-pldi-092_pdf](https://cloud.githubusercontent.com/assets/3731059/6526295/9187883a-c410-11e4-8b58-e695d813a006.jpg)

I did not totally hide the canvas, so you can see graphics:

![compressed_tracemonkey-pldi-091_pdf](https://cloud.githubusercontent.com/assets/3731059/6526305/ba9f4488-c410-11e4-96a7-62c22b7f3180.jpg)


But ultimately, #5545 is the solution!

**Comment by yurydelendik** [comment]
> one has to be really precise when selecting a text and not move mouse into the empty space between text because selection jumps to the beginning.

#6591 fixes "selection jumps to the beginning" part. Can we count this issue as resolved?

**Comment by yurydelendik** [comment]
>  joining all text segments you get the correct text content of the PDF file.

There is also #6590 to fix #1045

**Comment by speedplane** [comment]
Both #6590 and #6591 improve the situation with much less work, but they do not fix it. You still get jittery text selection, but it occurs in a less annoying way. I'm for fixing it altogether, but I understand that the maintainers have to consider long term maintainability of the code.

**Comment by timvandermeij** [comment]
Closing as fixed by #7539. We file a follow-up issue for enabling this by default and improving performance, but the problem as outlined here should be addressed now.
